bigtable/internal/testproxy/proxy.go

```diff
@@ -0,0 +1,976 @@
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO(developer): Determine how to pass this call option back to caller
+		option.WithTokenSource(c.TokenSource)
+
+		return nil, nil
+	}
+	return creds, nil
+}
+
+// goTestProxyServer represents an instance of the test proxy server. It keeps
+// a reference to individual clients instances (stored in a testClient object).
+type goTestProxyServer struct {
+	pb.UnimplementedCloudBigtableV2TestProxyServer
+	clientIDs   map[string]testClient // clientIDs has all of the bigtable.Client objects under test
+	clientsLock sync.Mutex
+}
+
+// CreateClient responds to the CreateClient RPC. This method adds a new client
+// instance to the goTestProxyServer
+func (s *goTestProxyServer) CreateClient(ctx context.Context, req *pb.CreateClientRequest) (*pb.CreateClientResponse, error) {
+	if req.ClientId == "" ||
+		req.DataTarget == "" ||
+		req.ProjectId == "" ||
+		req.InstanceId == "" {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s must provide ClientId, DataTarget, ProjectId, and InstanceId", logLabel))
+	}
+
+	s.clientsLock.Lock()
+	if _, exists := s.clientIDs[req.ClientId]; exists {
+		return nil, stat.Error(codes.AlreadyExists,
+			fmt.Sprintf("%s: ClientID already exists", logLabel))
+	}
+	s.clientsLock.Unlock()
+
+	opts, err := getCredentialsOptions(req)
+	if err != nil {
+		return nil, stat.Error(codes.Unauthenticated,
+			fmt.Sprintf("%s: failed to get credentials: %v", logLabel, err))
+	}
+
+	conn, err := grpc.Dial(req.DataTarget, opts...)
+	if err != nil {
+		return nil, stat.Error(codes.Unknown, fmt.Sprintf("%s: failed to create connection: %v", logLabel, err))
+	}
+
+	localCtx, cancel := context.WithCancel(context.Background())
+	c, err := bigtable.NewClient(localCtx, req.ProjectId, req.InstanceId, option.WithGRPCConn(conn))
+	if err != nil {
+		cancel()
+		return nil, stat.Error(codes.Internal,
+			fmt.Sprintf("%s: failed to create client: %v", logLabel, err))
+	}
+
+	if s.clientIDs == nil {
+		s.clientIDs = make(map[string]testClient)
+	}
+
+	s.clientsLock.Lock()
+	s.clientIDs[req.ClientId] = testClient{
+		c: c,
+		cancels: []context.CancelFunc{
+			cancel,
+		},
+		appProfileID:        req.AppProfileId,
+		perOperationTimeout: req.PerOperationTimeout,
+	}
+	s.clientsLock.Unlock()
+
+	return &pb.CreateClientResponse{}, nil
+}
+
+// RemoveClient responds to the RemoveClient RPC. This method removes an
+// existing client from the goTestProxyServer
+func (s *goTestProxyServer) RemoveClient(ctx context.Context, req *pb.RemoveClientRequest) (*pb.RemoveClientResponse, error) {
+	clientID := req.ClientId
+	doCancelAll := req.CancelAll
+
+	s.clientsLock.Lock()
+	btc, exists := s.clientIDs[clientID]
+	if !exists {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+	defer s.clientsLock.Unlock()
+
+	if doCancelAll {
+		btc.cancelAll()
+	}
+	btc.c.Close()
+	delete(s.clientIDs, clientID)
```

Comment:
This also needs to be protected.

Reply:
Is it covered by the deferred `Unlock()` above?